File: src/main/java/de/blazemcworld/fireflow/util/TextWidth.java
package de.blazemcworld.fireflow.util;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.format.TextDecoration;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class TextWidth {

    private static final HashMap<String, Info> known = new HashMap<>();
    private static final Info missing;

    static {
        try (InputStream stream = TextWidth.class.getClassLoader().getResourceAsStream("fontwidth.json")) {
            String raw = new String(stream.readAllBytes());

            JsonObject parsed = JsonParser.parseString(raw).getAsJsonObject();

            missing = new Gson().fromJson(parsed.get("missing_char"), Info.class);

            for (Map.Entry<String, JsonElement> entry : parsed.get("chars").getAsJsonObject().entrySet()) {
                known.put(entry.getKey(), new Gson().fromJson(entry.getValue(), Info.class));
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static double calculate(Component text) {
        return calculate(text, false);
    }

    public static double calculate(Component text, boolean bold) {
        bold = text.decoration(TextDecoration.BOLD) == TextDecoration.State.TRUE || (text.decoration(TextDecoration.BOLD) == TextDecoration.State.NOT_SET && bold);

        double width = 0;
        if (text instanceof TextComponent t) {
            width += calculate(t.content(), bold);
        }
        for (Component child : text.children()) {
            width += calculate(child, bold);
        }
        return width;
    }

    public static double calculate(String text, boolean bold) {
        double width = 0;

        for (String s : text.split("")) {
            Info info = known.getOrDefault(s, missing);
            width += info.width;
            if (bold) width += info.bold;
        }

        return width;
    }

    private record Info(double width, double bold) {}
}

File: src/main/kotlin/de/blazemcworld/fireflow/gui/TextWidth.kt
package de.blazemcworld.fireflow.gui

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.TextDecoration

object TextWidth {

    private val info: JsonObject
    private val missingWidth: Double
    private val missingBoldOffset: Double

    init {
        val data = JsonParser.parseString(TextWidth::class.java.getResource("/fontwidth.json")!!.readText()).asJsonObject
        info = data.get("chars").asJsonObject
        val missing = data.get("missing_char").asJsonObject
        missingWidth = missing.get("width").asDouble
        missingBoldOffset = missing.get("bold_offset").asDouble
    }

    fun calculate(comp: TextComponent, parentBold: Boolean = false): Double {
        val state = comp.style().decoration(TextDecoration.BOLD)
        var bold = parentBold
        if (state == TextDecoration.State.TRUE) bold = true
        if (state == TextDecoration.State.FALSE) bold = false

        var width = calculate(comp.content(), bold)
        for (child in comp.children()) {
            if (child is TextComponent) {
                width += calculate(child, bold)
            }
        }

        return width
    }

    fun calculate(str: String, bold: Boolean = false): Double {
        var width = 0.0

        for (c in str) {
            if (info.has(c.toString())) {
                width += info.get(c.toString()).asJsonObject.get("width").asDouble
                if (bold) width += info.get(c.toString()).asJsonObject.get("bold_offset").asDouble
            } else {
                width += missingWidth
                if (bold) width += missingBoldOffset
            }
        }

        return width
    }

}